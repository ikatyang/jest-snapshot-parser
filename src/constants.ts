// https://github.com/facebook/jest/blob/v20.0.4/packages/jest-snapshot/src/utils.js#L23
export const version_regex = /^\/\/ Jest Snapshot (v.+),/;

export type ParsedValue = undefined | null | boolean | number | string;

export interface Parsed {
  [key: string]: ParsedValue;
}
