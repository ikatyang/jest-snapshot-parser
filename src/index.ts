export * from './parse';
export * from './constants';
