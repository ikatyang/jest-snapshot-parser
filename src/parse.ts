import { version_regex, Parsed } from './constants';
import { parse_v1 } from './v1/parse';

export const parse = (content: string): Parsed => {
  const results = content.match(version_regex);

  if (results === null) {
    throw new Error(`Unexpected snapshot content, there is no snapshot header`);
  }

  const version = results[1];

  switch (version) {
    case 'v1':
      return parse_v1(content);
    default:
      throw new Error(`Unexpected snapshot version: ${version}`);
  }
};
