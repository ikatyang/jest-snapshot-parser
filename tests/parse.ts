import * as fs from 'fs';
import * as path from 'path';
import { parse } from '../src/index';

const loaded_baseline = load_baseline();

it('should throw error with invalid snapshot header', () => {
  expect(() => parse('')).toThrowErrorMatchingSnapshot();
});

it('should throw error with unexpected snapshot version', () => {
  expect(() => parse('// Jest Snapshot v0,')).toThrowErrorMatchingSnapshot();
});

it('should parse correctly with snapshot v1', () => {
  expect(parse(load_fixture('v1'))).toEqual(loaded_baseline);
});

function load_fixture(version: string) {
  return fs.readFileSync(
    path.resolve(__dirname, `../fixtures/${version}/test.js.snap`),
    'utf8',
  );
}

function load_baseline() {
  const data = require('../fixtures/data');
  const baseline: Record<string, string> = {};

  Object.keys(data).forEach(type => {
    const tests = data[type];
    Object.keys(tests).forEach(kind => {
      baseline[`${type} ${kind} 1`] = tests[kind];
    });
  });

  return baseline;
}
