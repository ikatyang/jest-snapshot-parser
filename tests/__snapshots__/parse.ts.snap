// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw error with invalid snapshot header 1`] = `"Unexpected snapshot content, there is no snapshot header"`;

exports[`should throw error with unexpected snapshot version 1`] = `"Unexpected snapshot version: v0"`;
